File: src/padflies/padflies/_padflie_tf.py
import rclpy
from rclpy.node import Node
import rclpy.duration
import rclpy.time
from rclpy.callback_groups import MutuallyExclusiveCallbackGroup
import tf2_py
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener
from geometry_msgs.msg import TransformStamped, PointStamped, PoseStamped, Quaternion
import tf2_geometry_msgs as tf2_geom
import tf_transformations

from crazyflie_interfaces_python.positions import CfPositionBuffer
from typing import Optional, Tuple, List, Callable


### This needs to be ported to crazyflie_interfaces_python
from rclpy.qos import QoSProfile, HistoryPolicy, DurabilityPolicy, ReliabilityPolicy
from crazyflie_interfaces.msg import PoseStampedArray


class CfPositionListener:
    def __init__(self, buffer: CfPositionBuffer, node: Node):
        self._buffer = buffer
        self._node = node
        self._callback_group = MutuallyExclusiveCallbackGroup()

        qos = QoSProfile(
            depth=100,
            reliability=ReliabilityPolicy.BEST_EFFORT,
            durability=DurabilityPolicy.VOLATILE,
            history=HistoryPolicy.KEEP_LAST,
        )

        self._cf_position_sub = node.create_subscription(
            msg_type=PoseStampedArray,
            topic="/cf_positions",
            callback=self.callback,
            qos_profile=qos,
            callback_group=self._callback_group,
        )

    def __del__(self) -> None:
        self.unregister()

    def unregister(self):
        self._node.destroy_subscription(self._cf_position_sub)

    def callback(self, msg: PoseStampedArray):
        pose_stamped: PoseStamped
        for pose_stamped in msg.poses:
            self._buffer.set_position(pose_stamped, msg.header.frame_id)


class PadflieTF:

    def __init__(
        self,
        node: Node,
        sleep: Callable[[float], None],
        pad_name: str,
        cf_name: str,
        world: str = "world",
    ):
        self._node = node
        self._sleep = sleep
        self._pad_name = pad_name
        self._cf_name = cf_name
        self._world = world

        self._tf_buffer = Buffer(cache_time=rclpy.duration.Duration(seconds=1))
        self._cf_buffer = CfPositionBuffer(node=node)

    def activate(self):
        self._tf_listener = TransformListener(self._tf_buffer, node=self._node)
        self._cf_listener = CfPositionListener(self._cf_buffer, node=self._node)

    def deactivate(self):
        self._tf_listener.unregister()
        self._cf_listener.unregister()

        del self._tf_listener
        del self._cf_listener

    def get_pad_position_or_timeout(
        self, timeout_sec: float
    ) -> Tuple[List[float], float]:
        """Retrieves the pad position or timess out after given time.

        Args:
            timeout_sec (float): The maximum time to wait for the pad position to become available.

        Raises:
            TimeoutError: If timeout occurs

        Returns:
            Tuple[List[float], float]: The position and yaw of the pad in world coordinates.
        """
        pad_position_and_yaw = self.get_pad_position_and_yaw()
        while pad_position_and_yaw is None and timeout_sec > 0.0:
            self._sleep(0.1)
            timeout_sec -= 0.1
            pad_position_and_yaw = self.get_pad_position_and_yaw()
        if pad_position_and_yaw is None:
            raise TimeoutError(f"Pad with name: {self._pad_name}, could not be found")
        return pad_position_and_yaw

    def get_pad_position_and_yaw(self) -> Optional[Tuple[List[float], float]]:
        """Gets the position of our pad.

        Returns:
            Optional[Tuple[List[float], float]]: None if not possible. Otherwise position and yaw of pad in world coords.
        """

        t = self.get_transform(
            target_frame=self._world,
            source_frame=self._pad_name,
        )
        if t is None:
            return None

        position = [
            t.transform.translation.x,
            t.transform.translation.y,
            t.transform.translation.z,
        ]
        quaternion = (
            t.transform.rotation.x,
            t.transform.rotation.y,
            t.transform.rotation.z,
            t.transform.rotation.w,
        )
        _roll, _pitch, yaw = tf_transformations.euler_from_quaternion(quaternion)
        return (position, yaw)

    def get_pad_pose(self) -> PoseStamped:
        pose = PoseStamped()
        pose.header.frame_id = self._pad_name
        return pose

    def get_pad_pose_world(self) -> Optional[PoseStamped]:
        transform = self.get_transform(
            target_frame=self._world,
            source_frame=self._pad_name,
        )
        if transform is None:
            return None
        zero_pose = PoseStamped()
        zero_pose.header.frame_id = self._world
        return tf2_geom.do_transform_pose_stamped(zero_pose, transform)

    def get_cf_pose_stamped(
        self,
        frame_id: Optional[str],
        world_quat: Quaternion = Quaternion(x=0.0, y=0.0, z=0.0, w=1.0),
    ) -> Optional[PoseStamped]:
        """Get the pose of the cf in a specifc frame_id
        Args:
            frame_id (Optional[str], optional): The frame to get the cf pose in. Defaults to None.
            world_quat (Quaternion, optional): A Quaternion which is added to the pose. Defaults to Quaternion(x=0.0, y=0.0, z=0.0, w=1.0).

        Returns:
            Optional[PoseStamped]: A Pose Stamped with frame_id as reference.
        """
        transform = self.get_transform(target_frame=frame_id, source_frame=self._world)
        world_pose = self._cf_buffer.get_position(self._cf_name)

        if transform is None or world_pose is None:
            return None
        return tf2_geom.do_transform_pose_stamped(world_pose, transform)

    def get_cf_position(self) -> Optional[List[float]]:
        """Gets the position of the crazyflie in world coordinates.

        Returns:
            Optional[List[float]]: The Position of the cf in world coordinates.
        """
        world_pose = self._cf_buffer.get_position(self._cf_name)
        if world_pose is None:
            return None
        return [
            world_pose.pose.position.x,
            world_pose.pose.position.y,
            world_pose.pose.position.z,
        ]

    def transform_point_stamped(
        self,
        point: PointStamped,
        target_frame: str,
    ) -> Optional[PointStamped]:
        """Transforms stamped point into another frame.

        The source frame is in the stamp of the point.

        Args:
            point (PointStamped): The point to transform (header frame_id is used as source frame).
            target_frame (str): The frame the point should get transformed to.
        Returns:
            Optional[PointStamped]: The transformed point. None if was not possible.
        """
        transform = self.get_transform(
            target_frame=target_frame, source_frame=point.header.frame_id
        )
        if transform is None:
            return None

        return tf2_geom.do_transform_point(point=point, transform=transform)

    def transform_pose_stamped(
        self, pose: PoseStamped, target_frame: str
    ) -> Optional[PoseStamped]:
        """Transforms a pose from the frame given in pose stamp to the target frame

        Similar to transform point stamped

        Args:
            pose (PoseStamped): The stamped pose to transform
            target_frame (str): The target frame to transform the pose to

        Returns:
            Optional[PoseStamped]: The new pose in the target frame
        """
        transform = self.get_transform(
            target_frame=target_frame, source_frame=pose.header.frame_id
        )
        if transform is None:
            return None

        return tf2_geom.do_transform_pose_stamped(pose=pose, transform=transform)

    def pose_stamped_to_world_position_and_yaw(
        self, pose: PoseStamped
    ) -> Optional[Tuple[List[float], float]]:
        world_pose = self.transform_pose_stamped(pose, self._world)
        if world_pose is None:
            return None
        _roll, _pitch, yaw = tf_transformations.euler_from_quaternion(
            (
                world_pose.pose.orientation.x,
                world_pose.pose.orientation.y,
                world_pose.pose.orientation.z,
                world_pose.pose.orientation.w,
            )
        )
        position = [
            world_pose.pose.position.x,
            world_pose.pose.position.y,
            world_pose.pose.position.z,
        ]
        return (position, yaw)

    def get_transform(
        self, target_frame: str, source_frame: str
    ) -> Optional[TransformStamped]:
        try:
            time = rclpy.time.Time(seconds=0, nanoseconds=0)
            # The core check needs to be done.
            # Otherwise the lookup transform timeout gets ignored
            if not self._tf_buffer.can_transform_core(target_frame, source_frame, time):
                return None

            return self._tf_buffer.lookup_transform_core(
                target_frame=target_frame,
                source_frame=source_frame,
                time=time,
            )

        except (tf2_py.LookupException, tf2_py.ConnectivityException) as ex:
            # This is most likely a Transform execption, which is expected and therefore None is ok.
            # It is very likely
            return None

File: src/padflies_cpp/src/padflie_tf.cpp
#include "padflies_cpp/padflie_tf.hpp"
#include <tf2/utils.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>

using std::placeholders::_1;


PadflieTF::PadflieTF(
    const std::string & cf_name,
    const std::string & world_frame)
: m_cf_name(cf_name) 
, m_world_frame(world_frame)
, m_has_pad(false)
, m_last_position()
, m_last_position_time(rclcpp::Time(0))
, m_position_timeout(rclcpp::Duration::from_seconds(1.0))
, m_tf_buffer(std::make_unique<tf2_ros::Buffer>(std::make_shared<rclcpp::Clock>(RCL_ROS_TIME)))
{
    RCLCPP_INFO(rclcpp::get_logger("TF"), "PadflieTF initialized for CF: %s, world frame: %s", 
                m_cf_name.c_str(), m_world_frame.c_str());      
}

PadflieTF::~PadflieTF()
{
    RCLCPP_INFO(rclcpp::get_logger("TF"), "PadflieTF destructor called");
}

void PadflieTF::on_configure(
    std::shared_ptr<rclcpp_lifecycle::LifecycleNode> node)
{
    m_node = node;
    m_tf_listener = std::make_unique<tf2_ros::TransformListener>(*m_tf_buffer, node);
    
    RCLCPP_INFO(node->get_logger(), "PadflieTF creating stufff");      

    m_cf_positions_sub = node->create_subscription<crazyflie_interfaces::msg::PoseStampedArray>(
       "/cf_positions",
       10, 
       std::bind(&PadflieTF::cf_positions_callback, this, _1));
}

void PadflieTF::set_pad(const std::string & pad_name)
{
    m_pad_name = pad_name;
    m_has_pad = true;
}

bool PadflieTF::get_pad_position_and_yaw_or_timeout(
    rclcpp::Duration & timeout_sec,
    Eigen::Vector3d & position,
    double & yaw)
{
    if (!m_has_pad) {
        log("Does not have a pad.");
        return false;
    }

    rclcpp::Time start_time = get_now();
    while (!get_pad_position_and_yaw(position, yaw)) {
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
        if (start_time + timeout_sec < get_now()) {
            log("Timeout while waiting for pad position.");
            return false;
        }
    }

    return true;
}

bool PadflieTF::get_pad_position_and_yaw(
    Eigen::Vector3d & position,
    double & yaw)
{
    if (!m_has_pad) {
        log("Does not have a pad.");
        return false;
    }

    geometry_msgs::msg::TransformStamped transform;
    if (lookup_transform(m_world_frame, m_pad_name, transform))
    {
        position = Eigen::Vector3d(
            transform.transform.translation.x,
            transform.transform.translation.y,
            transform.transform.translation.z);
        yaw = tf2::getYaw(transform.transform.rotation);
        return true;
    }
    return false;
}


bool PadflieTF::get_pad_pose(geometry_msgs::msg::PoseStamped & pose_stamped)
{
    if (!m_has_pad) {
        log("Does not have a pad.");
        return false;
    }
    pose_stamped = geometry_msgs::msg::PoseStamped();
    pose_stamped.header.frame_id = m_pad_name;
    return true;
}



bool PadflieTF::get_pad_pose_world(geometry_msgs::msg::PoseStamped & pose_stamped)
{
    if (!m_has_pad) {
        log("Does not have a pad.");
        return false;
    }

    geometry_msgs::msg::TransformStamped transform;

    if (lookup_transform(m_world_frame, m_pad_name, transform))
    {
        geometry_msgs::msg::PoseStamped zero_pose;
        zero_pose.header.frame_id = m_world_frame;
        tf2::doTransform(zero_pose, pose_stamped, transform);
        return true;
    }
    return false;
}

bool PadflieTF::get_cf_pose_stamped(
    const std::string & frame_id,
    geometry_msgs::msg::PoseStamped & pose_stamped)
{
    geometry_msgs::msg::TransformStamped transform;

    if (m_last_position_time + m_position_timeout > get_now()
        && lookup_transform(frame_id, m_world_frame, transform))
    {
        tf2::doTransform(m_last_position, pose_stamped, transform);    
        return true;
    }
    return false;
}


bool PadflieTF::get_cf_position(Eigen::Vector3d & position)
{
    if (m_last_position_time + m_position_timeout > get_now()) {
        position = Eigen::Vector3d(
            m_last_position.pose.position.x,
            m_last_position.pose.position.y,
            m_last_position.pose.position.z);
        return true;
    }
    return false;
}


bool PadflieTF::transform_point_stamped(
    const geometry_msgs::msg::PointStamped & point,
    const std::string & target_frame,
    geometry_msgs::msg::PointStamped & transformed_point)
{
    geometry_msgs::msg::TransformStamped transform;
    if (lookup_transform(target_frame, point.header.frame_id, transform)) {
        // Transform the point
        tf2::doTransform(point, transformed_point, transform);
        transformed_point.header.frame_id = target_frame;
        transformed_point.header.stamp = get_now();
        return true;
    }
    return false;
}

bool PadflieTF::transform_pose_stamped(
    const geometry_msgs::msg::PoseStamped & pose,
    const std::string & target_frame,
    geometry_msgs::msg::PoseStamped & transformed_pose)
{
    geometry_msgs::msg::TransformStamped transform;
    if (lookup_transform(target_frame, pose.header.frame_id, transform)) {
        // Transform the pose
        tf2::doTransform(pose, transformed_pose, transform);
        transformed_pose.header.frame_id = target_frame;
        transformed_pose.header.stamp = get_now();
        return true;
    }
    return false;
}

bool PadflieTF::pose_stamped_to_world_position_and_yaw(
    const geometry_msgs::msg::PoseStamped & pose_stamped,
    Eigen::Vector3d & position,
    double & yaw)
{
    geometry_msgs::msg::PoseStamped transformed_pose;
    if (transform_pose_stamped(pose_stamped, m_world_frame, transformed_pose)) 
    {
        position = Eigen::Vector3d(
            transformed_pose.pose.position.x,
            transformed_pose.pose.position.y,
            transformed_pose.pose.position.z);
        yaw = tf2::getYaw(transformed_pose.pose.orientation);
        return true;
    }
    return false;
}

bool PadflieTF::lookup_transform(
    const std::string & target_frame,
    const std::string & source_frame,
    geometry_msgs::msg::TransformStamped & transform)
{
    try {
        transform = m_tf_buffer->lookupTransform(target_frame, source_frame, rclcpp::Time(0));
        return true;
    } catch (const tf2::LookupException & ex) {
        log("LookupException: %s", ex.what());
    } catch (const tf2::ConnectivityException & ex) {
        log("ConnectivityException: %s", ex.what());
    } catch (const tf2::ExtrapolationException & ex) {
       log("ExtrapolationException: %s", ex.what());
    }

    return false;
}

void PadflieTF::cf_positions_callback(
    const crazyflie_interfaces::msg::PoseStampedArray::SharedPtr msg)
{
    for (const auto & pose : msg->poses) {
        if (pose.header.frame_id == m_cf_name) {
            m_last_position = pose;
            m_last_position_time = get_now();
        }
    }
}

template<typename... Args>
void PadflieTF::log(const char* format, Args&&... args)
{
    if (auto node_shared = m_node.lock()) {
        RCLCPP_INFO(node_shared->get_logger(), format, std::forward<Args>(args)...);
    }
}

rclcpp::Time PadflieTF::get_now() 
{
    static rclcpp::Time last_valid_time = rclcpp::Time(0);

    if (auto node_shared = m_node.lock()) {
        last_valid_time = node_shared->now();
    }
    return last_valid_time;
}